gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcDownscopingInterceptor.java

```diff
@@ -0,0 +1,226 @@
+
+/**
+ * Interceptor to set the downscoped token.
+ *
+ * <p>When downscoping is enabled, the downscoped token is generated for each request. Java storage
+ * client currently does not have the capability to set this for each of the requests. As a work
+ * around the credentials will be set from the interceptor.
+ *
+ * <p>This is a short term solution till Java storage client adds support for setting credentials at
+ * a request level.
+ */
+class GoogleCloudStorageClientGrpcDownscopingInterceptor implements ClientInterceptor {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_COMPOSE_OBJECT =
+      "google.storage.v2.Storage/ComposeObject";
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_DELETE_OBJECT =
+      "google.storage.v2.Storage/DeleteObject";
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_READ_OBJECT =
+      "google.storage.v2.Storage/ReadObject";
+;
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_START_RESUMABLE_WRITE =
+      "google.storage.v2.Storage/StartResumableWrite";
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_WRITE_OBJECT =
+      "google.storage.v2.Storage/WriteObject";
+
+  @VisibleForTesting private static final int BUCKET_PREFIX_LENGTH = "projects/_/buckets/".length();
+
+  @VisibleForTesting
+  static Metadata.Key<String> AUTH_KEY =
+      Metadata.Key.of(AUTHORIZATION, Metadata.ASCII_STRING_MARSHALLER);
+
+  private final Function<List<AccessBoundary>, String> downscopingFunction;
+
+  public GoogleCloudStorageClientGrpcDownscopingInterceptor(
+      Function<List<AccessBoundary>, String> downscopedAccessToken) {
+    this.downscopingFunction = downscopedAccessToken;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    logger.atFinest().log(String.format("interceptCall(): method=%s", method.getFullMethodName()));
+    return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+      private int flowControlRequests;
+      private String methodName;
+      private Metadata headers;
+      private Listener<RespT> responseListener;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        this.responseListener = responseListener;
+        this.headers = headers;
+        this.methodName = method.getFullMethodName();
+
+        logger.atFinest().log("start(): method=%s", methodName);
+      }
+
+      @Override
+      public void request(int numMessages) {
+        if (headers != null) {
+          this.flowControlRequests += numMessages;
+        } else {
+          super.request(numMessages);
+        }
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        setAuthHeader(message);
```

Comment:
is stream's header required to be updated for every request message? IIUC, it would only be functional if stream is not created yet.

Reply:
Auth header is updated only once since this.headers will be set to null.